stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/list/adapter/viewholder/internal/GiphyAttachmentViewHolder.kt

```diff
@@ -0,0 +1,133 @@
+package io.getstream.chat.android.ui.message.list.adapter.viewholder.internal
+
+import android.view.ViewGroup
+import androidx.constraintlayout.widget.ConstraintLayout
+import androidx.core.view.isVisible
+import androidx.core.view.updateLayoutParams
+import com.getstream.sdk.chat.adapter.MessageListItem
+import com.getstream.sdk.chat.utils.extensions.constrainViewEndToEndOfView
+import com.getstream.sdk.chat.utils.extensions.constrainViewStartToEndOfView
+import com.getstream.sdk.chat.utils.extensions.isBottomPosition
+import com.getstream.sdk.chat.utils.extensions.updateConstraints
+import io.getstream.chat.android.client.models.Attachment
+import io.getstream.chat.android.ui.common.extensions.hasText
+import io.getstream.chat.android.ui.common.extensions.internal.streamThemeInflater
+import io.getstream.chat.android.ui.common.extensions.isReply
+import io.getstream.chat.android.ui.databinding.StreamUiItemGiphyAttachmentBinding
+import io.getstream.chat.android.ui.message.list.adapter.MessageListItemPayloadDiff
+import io.getstream.chat.android.ui.message.list.adapter.MessageListListenerContainer
+import io.getstream.chat.android.ui.message.list.adapter.MessageListListenerContainerImpl
+import io.getstream.chat.android.ui.message.list.adapter.internal.DecoratedBaseMessageItemViewHolder
+import io.getstream.chat.android.ui.message.list.adapter.view.internal.GiphyMediaAttachmentView
+import io.getstream.chat.android.ui.message.list.adapter.viewholder.decorator.internal.BackgroundDecorator
+import io.getstream.chat.android.ui.message.list.adapter.viewholder.decorator.internal.Decorator
+import io.getstream.chat.android.ui.transformer.ChatMessageTextTransformer
+
+internal class GiphyAttachmentViewHolder(
+    parent: ViewGroup,
+    decorators: List<Decorator>,
+    private val listeners: MessageListListenerContainer?,
+    private val markdown: ChatMessageTextTransformer,
+    internal val binding: StreamUiItemGiphyAttachmentBinding = StreamUiItemGiphyAttachmentBinding.inflate(
+        parent.streamThemeInflater,
+        parent,
+        false
+    ),
+) : DecoratedBaseMessageItemViewHolder<MessageListItem.MessageItem>(binding.root, decorators) {
+
+    /**
+     * We override the Message passed to listeners here with the up-to-date Message
+     * object from the [data] property of the base ViewHolder.
+     *
+     * This is required because these listeners will be invoked by the AttachmentViews,
+     * which don't always have an up-to-date Message object in them. This is due to the
+     * optimization that we don't re-create the AttachmentViews when the attachments
+     * of the Message are unchanged. However, other properties (like reactions) might
+     * change, and these listeners should receive a fully up-to-date Message.
+     */
+    private fun modifiedListeners(listeners: MessageListListenerContainer?): MessageListListenerContainer? {
```

Comment:
Do we need this part because it seems to me that this is a use case specific to `ViewHolder`s holding `AttachmentViewFactory`?

Reply:
This is needed because the content of `data` maybe change. There's no guarantee that the `data.message` that was used in the `MessageListListenerContainer` at the moment of the creation of this ViewHolder will be update-to-date when a call happens in one of the listeners, so we need to update the listeners of every `bind()`.

Example: 
- This view holder is created and `MessageListListenerContainer` is passed in its constructor. 
- A reaction is added to the message of this view holder. 
- `public fun onMessageClick(message: Message)` of `listeners` will provide a message without the added reaction in its callback, because `listeners` wasn't updated.

Reply:
Thanks Leo!